File: ts/src/utils/serialize.ts
/** Recursive merge utility and UNSET sentinel.
 * Ported from src/minisweagent/utils/serialize.py */

export const UNSET = Symbol("UNSET");

export type MaybeUnset<T> = T | typeof UNSET;

/** Merge multiple dictionaries recursively.
 * Later dictionaries take precedence over earlier ones.
 * Nested dictionaries are merged recursively.
 * UNSET values are skipped. */
export function recursiveMerge(...dicts: (Record<string, unknown> | null | undefined)[]): Record<string, unknown> {
  if (dicts.length === 0) return {};
  const result: Record<string, unknown> = {};
  for (const d of dicts) {
    if (d == null) continue;
    for (const [key, value] of Object.entries(d)) {
      if (value === UNSET) continue;
      if (key in result && isPlainObject(result[key]) && isPlainObject(value)) {
        result[key] = recursiveMerge(result[key] as Record<string, unknown>, value as Record<string, unknown>);
      } else if (isPlainObject(value)) {
        // Recursively merge dict values to filter out nested UNSET values
        result[key] = recursiveMerge(value as Record<string, unknown>);
      } else {
        result[key] = value;
      }
    }
  }
  return result;
}

function isPlainObject(v: unknown): v is Record<string, unknown> {
  return typeof v === "object" && v !== null && !Array.isArray(v);
}

File: go/internal/serialize/serialize.go
// Package serialize provides recursive merge and UNSET sentinel.
// Ported from src/minisweagent/utils/serialize.py
package serialize

// Unset is a sentinel value that is skipped during recursive merge.
type UnsetType struct{}

var Unset = UnsetType{}

// IsUnset returns true if the value is the Unset sentinel.
func IsUnset(v any) bool {
	_, ok := v.(UnsetType)
	return ok
}

// UnsetIfEmpty returns Unset if s is empty, otherwise returns s.
func UnsetIfEmpty(s string) any {
	if s == "" {
		return Unset
	}
	return s
}

// RecursiveMerge merges multiple maps recursively.
// Later maps take precedence over earlier ones.
// Nested maps are merged recursively.
// Unset values are skipped.
func RecursiveMerge(dicts ...map[string]any) map[string]any {
	if len(dicts) == 0 {
		return map[string]any{}
	}
	result := map[string]any{}
	for _, d := range dicts {
		if d == nil {
			continue
		}
		for key, value := range d {
			if IsUnset(value) {
				continue
			}
			if existing, ok := result[key]; ok {
				if existMap, ok1 := existing.(map[string]any); ok1 {
					if valMap, ok2 := value.(map[string]any); ok2 {
						result[key] = RecursiveMerge(existMap, valMap)
						continue
					}
				}
			}
			if valMap, ok := value.(map[string]any); ok {
				// Recursively merge to filter out nested Unset values
				result[key] = RecursiveMerge(valMap)
			} else {
				result[key] = value
			}
		}
	}
	return result
}
